### PR TITLE
Fix callbacks

### LINKED ISSUE #758: Generates invalid callback wrappers with GCab
gir generates invalid code with gcab extract() functions (https://github.com/elmarco/gcab-rs). They are peculiar, they have 2 (synchronous) callbacks.

- code doesn't compile 
- callback is called without destructuring the tuple
- the second callback uses gio::Option (instead of presumably just Option)

```rust
    fn extract<P: IsA<gio::File>, Q: IsA<gio::Cancellable>>(&self, path: Option<&P>, file_callback: Option<&mut dyn (FnMut(&File) -> bool)>, progress_callback: gio::Option<&mut dyn (FnMut(i64, i64))>, cancellable: Option<&Q>) -> Result<(), Error> {
        let file_callback_data: &Option<&mut dyn (FnMut(&File) -> bool)> = &file_callback;
        unsafe extern "C" fn file_callback_func<P: IsA<gio::File>, Q: IsA<gio::Cancellable>>(file: *mut gcab_sys::GCabFile, user_data: glib_sys::gpointer) -> glib_sys::gboolean {
            let file = from_glib_borrow(file);
            let callback: Box_<(Option<&mut dyn (FnMut(&File) -> bool)>, gio::Option<&mut dyn (FnMut(i64, i64))>)> = *(user_data as *mut _);
            let res = 
                callback(&file)
            } else {
                panic!("cannot get closure...")
            };
            res.to_glib()
        }
        let file_callback = if file_callback_data.is_some() { Some(file_callback_func::<P, Q> as _) } else { None };
        let progress_callback_data: &gio::Option<&mut dyn (FnMut(i64, i64))> = &progress_callback;
        unsafe extern "C" fn progress_callback_func<P: IsA<gio::File>, Q: IsA<gio::Cancellable>>(current_num_bytes: i64, total_num_bytes: i64, user_data: glib_sys::gpointer) {
            let callback: Box_<(Option<&mut dyn (FnMut(&File) -> bool)>, gio::Option<&mut dyn (FnMut(i64, i64))>)> = *(user_data as *mut _);
                callback(current_num_bytes, total_num_bytes)
            } else {
                panic!("cannot get closure...")
            };
        }
        let progress_callback = if progress_callback_data.is_some() { Some(progress_callback_func::<P, Q> as _) } else { None };
        let super_callback0: Box_<(Option<&mut dyn (FnMut(&File) -> bool)>, gio::Option<&mut dyn (FnMut(i64, i64))>)> = Box_::new(Box_::new((file_callback_data, progress_callback_data)));
        unsafe {
            let mut error = ptr::null_mut();
            let _ = gcab_sys::gcab_cabinet_extract(self.as_ref().to_glib_none().0, path.map(|p| p.as_ref()).to_glib_none().0, file_callback, progress_callback, super_callback0 as *const _ as usize as *mut _, cancellable.map(|p| p.as_ref()).to_glib_none().0, &mut error);
            if error.is_null() { Ok(()) } else { Err(from_glib_full(error)) }
        }
    }

```

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #758.

cc @EPashkin @sdroege @elmarco

### CODE REVIEW

**Comment by sdroege** [comment]
Does it change anything in any of our crates?

**Comment by GuillaumeGomez** [comment]
It changes the version files yes.

**Comment by sdroege** [comment]
Ok, fine then ;)

**Comment by GuillaumeGomez** [comment]
The `if` condition isn't generated apparently. Needs to be fixed as well.

**Comment by EPashkin** [comment]
PR adds many functions like `threads_add_idle` to gdk,
SubprocessLauncher::set_child_setup to gio,
result compiles on Windows.


```rust
pub fn threads_add_idle<P: Fn() -> bool + Send + Sync + 'static>(function: P) -> u32 {
    assert_initialized_main_thread!();
    let function_data: Box_<P> = Box::new(function);
    unsafe extern "C" fn function_func<P: Fn() -> bool + Send + Sync + 'static>(user_data: glib_sys::gpointer) -> glib_sys::gboolean {
        let callback: &P = &*(user_data as *mut _);
        let res = (*callback)();
        res.to_glib()
    }
    let function = Some(function_func::<P> as _);
    let super_callback0: Box_<P> = function_data;
    unsafe {
        gdk_sys::gdk_threads_add_idle(function, Box::into_raw(super_callback0) as *mut _)
    }
}

    #[cfg(any(unix, feature = "dox"))]
    pub fn set_child_setup<P: Fn() + 'static>(&self, child_setup: P) {
        let child_setup_data: Box_<P> = Box::new(child_setup);
        unsafe extern "C" fn child_setup_func<P: Fn() + 'static>(user_data: glib_sys::gpointer) {
            let callback: &P = &*(user_data as *mut _);
            (*callback)();
        }
        let child_setup = Some(child_setup_func::<P> as _);
        unsafe extern "C" fn destroy_notify_func<P: Fn() + 'static>(data: glib_sys::gpointer) {
            let _callback: Box_<P> = Box_::from_raw(data as *mut _);
        }
        let destroy_call3 = Some(destroy_notify_func::<P> as _);
        let super_callback0: Box_<P> = child_setup_data;
        unsafe {
            gio_sys::g_subprocess_launcher_set_child_setup(self.to_glib_none().0, child_setup, Box::into_raw(super_callback0) as *mut _, destroy_call3);
        }
    }
```

File: src/codegen/function_body_chunk.rs [hunk 1]
```diff
@@ -348,15 +360,24 @@ impl Builder {
...
+                                format!("{}if let Some(ref callback) = callback{} {{",
                                         if trampoline.ret.c_type != "void" {
                                             "let res = "
                                         } else {
                                             ""
                                         },
-                                        pos)));
+                                        if let Some(pos) = pos {
+                                        format!(".{}", pos)
+                                    } else {
```
Comment (EPashkin) [inline src/codegen/function_body_chunk.rs]: unnecessary tab before `if let`

**Comment by EPashkin** [comment]
@GuillaumeGomez Thanks, LGFM

**Comment by GuillaumeGomez** [comment]
No no! It's not ready yet!

File: src/codegen/function_body_chunk.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/codegen/function_body_chunk.rs]: The other way around actually: missing tab for this line, the one before and the one after. :p

**Comment by GuillaumeGomez** [comment]
That \*should\* be good now. Just waiting for @elmarco confirmation.

**Comment by GuillaumeGomez** [comment]
Ah hold on, new error spotted!

**Comment by GuillaumeGomez** [comment]
Finally ready!

**Comment by GuillaumeGomez** [comment]
It generates code like this:

```rust
fn extract<P: IsA<gio::File>, Q: IsA<gio::Cancellable>>(&self, path: Option<&P>, file_callback: Option<&mut dyn (FnMut(&File) -> bool)>, progress_callback: Option<&mut dyn (FnMut(i64, i64))>, cancellable: Option<&Q>) -> Result<(), Error> {
    let file_callback_data: &Option<&mut dyn (FnMut(&File) -> bool)> = &file_callback;
    unsafe extern "C" fn file_callback_func<P: IsA<gio::File>, Q: IsA<gio::Cancellable>>(file: *mut gcab_sys::GCabFile, user_data: glib_sys::gpointer) -> glib_sys::gboolean {
        let file = from_glib_borrow(file);
        let callback: &mut (&mut Option<&mut dyn (FnMut(&File) -> bool)>, &mut Option<&mut dyn (FnMut(i64, i64))>) = &mut *(user_data as *mut _);
        let res = if let Some(ref mut callback) = (*callback).0 {
            callback(&file)
        } else {
            panic!("cannot get closure...")
        };
        res.to_glib()
    }
    let file_callback = if file_callback_data.is_some() { Some(file_callback_func::<P, Q> as _) } else { None };
    let progress_callback_data: &Option<&mut dyn (FnMut(i64, i64))> = &progress_callback;
    unsafe extern "C" fn progress_callback_func<P: IsA<gio::File>, Q: IsA<gio::Cancellable>>(current_num_bytes: i64, total_num_bytes: i64, user_data: glib_sys::gpointer) {
        let callback: &mut (&mut Option<&mut dyn (FnMut(&File) -> bool)>, &mut Option<&mut dyn (FnMut(i64, i64))>) = &mut *(user_data as *mut _);
        if let Some(ref mut callback) = (*callback).1 {
            callback(current_num_bytes, total_num_bytes)
        } else {
            panic!("cannot get closure...")
        };
    }
    let progress_callback = if progress_callback_data.is_some() { Some(progress_callback_func::<P, Q> as _) } else { None };
    let super_callback0: &(&Option<&mut dyn (FnMut(&File) -> bool)>, &Option<&mut dyn (FnMut(i64, i64))>) = &(file_callback_data, progress_callback_data);
    unsafe {
        let mut error = ptr::null_mut();
        let _ = gcab_sys::gcab_cabinet_extract(self.as_ref().to_glib_none().0, path.map(|p| p.as_ref()).to_glib_none().0, file_callback, progress_callback, super_callback0 as *const _ as usize as *mut _, cancellable.map(|p| p.as_ref()).to_glib_none().0, &mut error);
        if error.is_null() { Ok(()) } else { Err(from_glib_full(error)) }
    }
}
```

**Comment by EPashkin** [comment]
@GuillaumeGomez Thanks, looks good, and changes in gdk/gio still same
You want me merge after CI, or still waiting for @elmarco confirmation?

**Comment by GuillaumeGomez** [comment]
@elmarco confirmed (he's sitting next to me currently, that helps ;) )

**Comment by GuillaumeGomez** [comment]
But let's wait for CI in any case!

**Comment by GuillaumeGomez** [comment]
All green.

**Comment by EPashkin** [comment]
Thanks again

**Pull request merged**